### PR TITLE
Add missing decompositons/lowerings for logical/bitwise operators

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102566

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel

### CODE REVIEW

**Comment by alexsamardzic** [comment]
Following code would generate `Using FallbackKernel: torch.ops.aten.__lshift__.Scalar` warning on compilation:
```python
import torch

@torch.compile
def foo(x, y):
    return x | (y << 1)

if __name__ == '__main__':
    m, n = 3, 5
    device = 'cuda'

    x = torch.rand((m, n), device=device) < 0.5
    y = torch.rand((m, n), device=device) < 0.5
    
    print(foo(x, y))
```
This change fixes it.

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -3434,6 +3433,15 @@ def arange_start(
...
 
+@register_decomposition(aten.__lshift__.Scalar)
+def lshift_Scalar(self, value):
+    return torch.bitwise_left_shift(self, value)
+
+@register_decomposition(aten.__lshift__.Tensor)
+def lshift_Tensor(self, value):
+    return torch.bitwise_left_shift(self, value)
+
+
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:3444]:
```suggestion
@register_decomposition(aten.__lshift__)
def lshift(self, value):
    return torch.bitwise_left_shift(self, value)
```

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -155,7 +155,6 @@ def fill_tensor(self, value: Tensor):
     )
     return torch.full_like(self, value.item())
 
-
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:158]: leftover

File: torch/_decomp/__init__.py [hunk 3]
```diff
@@ -255,6 +255,8 @@ def core_aten_decompositions() -> Dict[OpOverload, Callable]:
             aten._log_softmax_backward_data,
             aten.logspace,
             aten.logsumexp.default,
+            aten.__lshift__.Scalar,
+            aten.__lshift__.Tensor,
```
Comment (lezcano) [inline torch/_decomp/__init__.py:259]: You can probably just write
```suggestion
            aten.__lshift__,
```
to select them all.

**Review by lezcano — COMMENTED** [review]
This is morally correct, but it can be golfed through some sugar that these decorators have.

Also, this seems like it's fixing some tests, so could you add the couple lines necessary to implement the other dunder logical operators in terms of the `torch.*` variants. With that, you should be able to remove the lowerings for `__{and,or,xor}__` in `lowering.py`.

Regarding testing, I'd love to say we have a simply way to add testing for this in terms of `torch.bitwise_left_shift`, but we don't (ugh). A proper fix would account for replacing the `alias` kwarg in OpInfos to, rather than take a string, take a function. That way you could add `Tensor.__lshift__` to the aliases of `torch.bitwise_left_shift`. Now, this may take some time, so better skip it for now. I'll open an issue discussing this in case anyone from OSS wants to pick it up.

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -3434,6 +3434,26 @@ def arange_start(
...
+    return torch.bitwise_or(self, value)
+
+
+@register_decomposition(aten.__lshift__)
+def dunder_lshift(self, value):
+    return torch.bitwise_left_shift(self, value)
+
+
+@register_decomposition(aten.__rshift__)
+def dunder_rshift(self, value):
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:3453]: Given that they have `out` variants, decorate them with the `out_wrapper` decorator.

**Comment by alexsamardzic** [comment]
Applied all suggestions from above, added the same for `__rshift__`, `__and__` and `__or__` operators.  Silenced linter, and fixed some tests.

The code generation for `__and__` operator was not working consistently: it would put `&&` there for CPU, and `&` for Triton.  Thus the changes in `triton.py`, and then bitwise ops are consistenly applied in lowering.py.  I think `logical_and` and `logical_or` ops should be changed in `triton.py` too, to have `&&` and `||` emitted instead.

**Comment by lezcano** [comment]
Even more, they should be changed for `and` and `or`! It'd be a nice short follow-up PR that you can stack on top of this one: Have a look at the semantics of `and` (or `&&`) in triton and fix `logical_and` and `logical_or`.
You will also need to change a few places where masks are being concatenated via a `" & ".join(masks)` and so. In general, it'd be good to audit all the `&` and `|` symbols inside the `_inductor` folder.

File: torch/_decomp/decompositions.py (hunk 4 above)
Comment (alexsamardzic) [reply torch/_decomp/decompositions.py:3457]: Fixed.

**Review by lezcano — APPROVED** [review]

**Comment by alexsamardzic** [comment]
Want to do `__xor__`, and maybe comparison operators too, before merging.  But need to look into semantics of corresponding ATen operators in more detail, it's confusing as for example there are both `__and__` and `bitwise_and` there...

**Review by alexsio27444 — APPROVED** [review]

File: torch/_inductor/lowering.py [hunk 5]
```diff
@@ -3868,16 +3867,56 @@ def truncdiv(a, b):
...
+
+
+@make_pointwise
+def bitwise_left_shift(a, b):
+    return ops.bitwise_left_shift(a, b)
+
+
+@make_pointwise
+def bitwise_right_shift(a, b):
+    return ops.bitwise_right_shift(a, b)
```
Comment (lezcano) [inline torch/_inductor/lowering.py:3917]: You can simply assign to variables the result of `register_pointwise` below and then reuse them to register the in-place ones.

Comment (alexsamardzic) [reply torch/_inductor/lowering.py]: Fixed.

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -296,14 +296,46 @@ def log10(x):
...
+        return f"{a} ^ 1"  # !{a} is not recognized by Triton
+
     @staticmethod
     def logical_and(a, b):
-        return f"{a} & {b}"
+        return f"{a} & {b}"  # {a} && {b} is not recognized by Triton
 
     @staticmethod
     def logical_or(a, b):
+        return f"{a} | {b}"  # {a} || {b} is not recognized by Triton
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:309]: Try `{a} and {b}`. That one works and I believe it does what you want. Similar for not, and etc.

File: torch/_inductor/codegen/triton.py [hunk 7]
```diff
@@ -296,14 +296,46 @@ def log10(x):
...
+
+    @staticmethod
+    def logical_xor(a, b):
+        return f"{a} ^ {b}"
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:313]: Also, due to precedence, you may need to put parens on these? I believe that there's a function to conditionally put parens on an expression if it doesn't have it yet.

File: torch/_inductor/codegen/triton.py (hunk 6 above)
Comment (alexsamardzic) [reply torch/_inductor/codegen/triton.py]: Fixed.

**Comment by alexsamardzic** [comment]
This is ready for review now.  Some comments:

- There are no parens around other expressions in corresponding strings produced by codegen (for example, `add` method produces `f"{a} + {b}"` in `cpp.py` codegen), so I haven't put any in my changes.
- As far as using `&` or `&&` C++ operators concerned for `logical_and` lowerings (and alike for `logical_or`), I realized later there were two versions of these lowerings for C++ codegen: one in `CppVecOverrides` and the other one in `CppOverrides` class.  Indeed it seems that in the first case, for some reason, operands are not converted to booleans (it seems the same is the case for Triton kernels generated), so an implementation with comparing them to 0, and then doing `&`, is more appropriate.  So I've added the rest of the logical operators on both places, but there are some things to investigate further, maybe in follow-up PR(s):
    - Should the inputs be converted to booleans in mentioned cases where it doesn't happen?
    - The code under `torch/_inductor` should be audited for uses of these C++ operators (like `" & ".join(masks)` that @lezcano mentioned).
    - Vectorized code is not actually generated for these operators, even if there exist lowerings in `CppVecOverrides` class, but it's generated instead if these are under some other vectorized expression, like `torch.where` in `test_vec_logical` test, in `test/inductor/test_cpu_repro.py` - that was the case also before my changes, not sure is it as intended, or something to be further investigated?
- Was using following simple test to verify that results of compiled functions match results of eager execution.  Not sure do we have/want alike test under `test/inductor`?

<details>
<summary>
Test to validate logical/bitwise operators compositions/lowering
</summary>

```Python
import torch


def make_fn1(fn):
    return lambda x: fn(x)


def make_fn2(fn):
    return lambda x, y: fn(x, y)


# torch.manual_seed(1)

m, n = 30, 40
device = "cpu"
#device = "cuda"

x = torch.randint(0, 10, (m, n), device=device)
y = torch.randint(0, 10, (m, n), device=device)
xb = x > 5
yb = y > 5

for op in (torch.logical_not, torch.bitwise_not):
    fn = make_fn1(op)
    fn_comp = torch.compile(fn)

    assert torch.all(fn(x) == fn_comp(x))
    assert torch.all(fn(xb) == fn_comp(xb))

for op in (
    torch.logical_and,
    torch.logical_or,
    torch.logical_xor,
    torch.bitwise_and,
    torch.bitwise_or,
    torch.bitwise_xor,
    torch.bitwise_left_shift,
    torch.bitwise_right_shift,
    torch.ops.aten.__and__,
    torch.ops.aten.__lshift__,
    torch.ops.aten.__or__,
    torch.ops.aten.__rshift__,
    torch.ops.aten.__xor__,
):
    fn = make_fn2(op)
    fn_comp = torch.compile(fn)

    assert torch.all(fn(x, y) == fn_comp(x, y))
    if not op in (
        torch.bitwise_left_shift,
        torch.bitwise_right_shift,
        torch.ops.aten.__lshift__,
        torch.ops.aten.__rshift__,
    ):
        assert torch.all(fn(xb, yb) == fn_comp(xb, yb))
```
</details>

**Review by lezcano — COMMENTED** [review]
Ok, this looks good. Now, I'm a bit concerned about moving from `a & b` in triton to `(a != 0) and (b != 0)`.

To see how does this affect performance, trigger a perf benchmark in
https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-compare.yml
by running the workflow on your branch. Then, you can see the results in
https://hud.pytorch.org/benchmark/compilers
and see if there's any regression.

An even better way of doing this would be to look at the generated triton IR / LLVM lowering and see if they generate the same code.

**Comment by lezcano** [comment]
@jgong5 see the follow-up items on the C++ vectorisation side.
https://github.com/pytorch/pytorch/pull/102566#issuecomment-1573367277

**Comment by alexsamardzic** [comment]
> Ok, this looks good. Now, I'm a bit concerned about moving from `a & b` in triton to `(a != 0) and (b != 0)`.
> 

Correct, the former is faster - I made the change for all Triton lowerings for logical operators, it's running on CI now.

My claim in the previous comment that `convert_input_to_bool=True` doesn't work for Triton was wrong.  I was thinking this as if `f"{a} and {b}"` lowering used for `logical_and`, in some cases the Triton code generated was not syntactically correct.  But that was for `torch.bool` inputs only, because `torch.bool` is actually 8-bit datatype, and Triton allow for `and` operator only over Boolean operands, so here the Triton expression generated, like `tmp4 = tmp2 and tmp3`, was not syntactically correct.  For non-`torch.bool` input, `convert_input_to_bool=True` would produce conversions `tmp4 = tmp2 != 0` and `tmp5 = tmp3 != 0` in Triton, and then an expression `tmp6 = tmp4 and tmp5` would be syntactically correct.  Thus, `and`/`or` cannot be used in Triton lowerings, and bitwise operators are better instead.

**Review by lezcano — APPROVED** [review]
Makes sense. Thanks for the detailed analysis!

**Review by jgong5 — APPROVED** [review]
LGTM for cpp related changes.
